internal/metrics/metrics.go

```diff
@@ -27,40 +40,61 @@ const (
 // desired endpoint. It should be invoked on application shutdown.
 type ReportFunc func()
 
-// MeterProvider returns a MeterProvider suitable for CLI usage.
+// Initialize returns a context.Context with a MeterProvider suitable for CLI usage.
 // The primary difference between this metric reader and a more typical
 // usage is that metric reporting only happens once when ReportFunc
 // is invoked.
-func MeterProvider(cli command.Cli) (metric.MeterProvider, ReportFunc, error) {
+func Initialize(ctx context.Context, cli command.Cli) (context.Context, ReportFunc, error) {
 	var exps []sdkmetric.Exporter
 
-	if exp, err := dockerOtelExporter(cli); err != nil {
-		return nil, nil, err
-	} else if exp != nil {
-		exps = append(exps, exp)
-	}
+	// Only metric exporters if the experimental flag is set.
+	if env.IsExperimental() {
+		if exp, err := dockerOtelExporter(cli); err != nil {
+			return nil, nil, err
+		} else if exp != nil {
+			exps = append(exps, exp)
+		}
 
-	if exp, err := detectOtlpExporter(context.Background()); err != nil {
-		return nil, nil, err
-	} else if exp != nil {
-		exps = append(exps, exp)
+		if exp, err := detectOtlpExporter(context.Background()); err != nil {
+			return nil, nil, err
+		} else if exp != nil {
+			exps = append(exps, exp)
+		}
 	}
 
 	if len(exps) == 0 {
 		// No exporters are configured so use a noop provider.
-		return noop.NewMeterProvider(), func() {}, nil
+		return ctx, func() {}, nil
 	}
 
-	// Use delta temporality because, since this is a CLI program, we can never
-	// know the cumulative value.
 	reader := sdkmetric.NewManualReader(
 		sdkmetric.WithTemporalitySelector(deltaTemporality),
 	)
 	mp := sdkmetric.NewMeterProvider(
-		sdkmetric.WithResource(detect.Resource()),
+		sdkmetric.WithResource(Resource()),
 		sdkmetric.WithReader(reader),
 	)
-	return mp, reportFunc(reader, exps), nil
+	return withMeterProvider(ctx, mp), reportFunc(reader, exps), nil
+}
+
+func withMeterProvider(ctx context.Context, mp metric.MeterProvider) context.Context {
+	return context.WithValue(ctx, meterProviderKey, mp)
```

Comment:
It looks like this is just global and there is no state in the context that changes metrics (unlike tracing). In that case, I think we should just pass the meterprovider (or I'm ok with global one as well if it is initialized from `cmd`, or associated with a specific package, eg. this function can just return it if you want). Context shouldn't be used just to hide parameters that functions take.

This can also easily break with nobody noticing that some callpath did not initialize context.

Reply:
Removed in favor of just passing it through.